[JSC] Improve BlockOrder by handling finished chains better and optimize specific common block structures.

h4writer opened the pull request:
#### 01f7679257a549dc02da742088490030c88cd41b
<pre>
[JSC] Improve BlockOrder by handling finished chains better and optimize specific common block structures.
<a href="https://bugs.webkit.org/show_bug.cgi?id=301564">https://bugs.webkit.org/show_bug.cgi?id=301564</a>

Reviewed by Yusuke Suzuki.

This improves the block order.

1) When there is no successor and the chain is broken,
take a better candidate using FIFO instead of LIFO.
2% more fall-through + 4% decrease in jump lengths

2) Detect common structures and put them in a better order.
Support is included for triangle, diamond, and
two successors (with one having other predecessors).
2.7% more fall-through + 11% decrease in jump lengths

* Source/JavaScriptCore/b3/air/AirOptimizeBlockOrder.cpp:
(JSC::B3::Air::detectTriangleStructure):
(JSC::B3::Air::detectDiamondStructure):
(JSC::B3::Air::detectExclusiveSuccessor):

Canonical link: <a href="https://commits.webkit.org/302875@main">https://commits.webkit.org/302875@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2dfc6bac1256b2d878d0d27b0d7a77e3fa3cd270

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130141 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2412 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41095 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137541 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81685 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2384 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2300 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99145 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66957 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133088 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1794 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116566 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79839 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1712 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34694 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80815 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/122144 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110228 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35202 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140020 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/128593 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2202 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2057 "Found 1 new test failure: imported/w3c/web-platform-tests/webvtt/rendering/cues-with-video/processing-model/align_center.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107668 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2246 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112910 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107546 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1754 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31360 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/55106 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20328 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2272 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65659 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/161608 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2089 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40294 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2293 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2198 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/b3/air/AirOptimizeBlockOrder.cpp:
```diff
@@ -57,29 +134,190 @@ class SortedSuccessors {
+}; // anonymous namespace
+
+static bool
+detectTriangleStructure(BasicBlock* A, ChainWorklist &worklist, IndexSet<BasicBlock*> &done)
```
This should be one line to follow to WebKit coding style.

Constellation commented on Source/JavaScriptCore/b3/air/AirOptimizeBlockOrder.cpp:
```diff
@@ -57,29 +134,190 @@ class SortedSuccessors {
 };
 
-} // anonymous namespace
+}; // anonymous namespace
```
Remove this semicolon.

Constellation commented on Source/JavaScriptCore/b3/air/AirOptimizeBlockOrder.cpp:
```diff
@@ -57,29 +134,190 @@ class SortedSuccessors {
+}
+
+static bool
+detectDiamondStructure(BasicBlock* B, ChainWorklist &worklist, IndexSet<BasicBlock*> &done)
```
Ditto.

Constellation commented on Source/JavaScriptCore/b3/air/AirOptimizeBlockOrder.cpp:
```diff
@@ -57,29 +134,190 @@ class SortedSuccessors {
+}
+
+static bool
+optimizeDualSuccessors(BasicBlock* block, ChainWorklist &worklist, IndexSet<BasicBlock*> &done)
```
Ditto.

Constellation reviewed (commented):
Looks cool!! JCS -> JSC. I've left some drive-by comments, but will look deeply.

Constellation commented on Source/JavaScriptCore/b3/air/AirOptimizeBlockOrder.cpp:
```diff
@@ -72,14 +166,155 @@ class SortedSuccessors {
 
 } // anonymous namespace
 
+static bool detectTriangleStructure(BasicBlock* A, ChainWorklist &worklist, IndexSet<BasicBlock*> &done)
```
Let's have
1. Naming `BasicBlock* A` to `BasicBlock* blockA` etc. WebKit style requires that variable is always starting with lowercase. Maybe `a` would be fine, but I feel `blockA` is good for readability & grep-ability
2. Let's make `&` attached to the typename side. `ChainWorklist& worklist`. Ditto for `done`.

Constellation commented on Source/JavaScriptCore/b3/air/AirOptimizeBlockOrder.cpp:
```diff
@@ -72,14 +166,155 @@ class SortedSuccessors {
+        if (!done.contains(C))
+            worklist.append(C);
+        return true;
+    }
```
These two cases look exactly the same, so how about doing it,

```
auto attemptToDetect = [&](BasicBlock* blockB, BasicBlock* blockC) {
    if ((blockC->numSuccessors() >= 1 && blockC->successor(0).block() == blockB)
        || (blockC->numSuccessors() >= 2 && blockC->successor(1).block() == blockB)) {
        if (!done.contains(blockB))
            worklist.append(blockB);
        if (!done.contains(blockC))
            worklist.append(blockC);
        return true;
    }
    return false;
};

auto* block0 = blockA->successor(0).block();
auto* block1 = blockA->successor(1).block();
return attemptToDetect(block0, block1) || attemptToDetect(block1, block0);

Constellation commented on Source/JavaScriptCore/b3/air/AirOptimizeBlockOrder.cpp:
```diff
@@ -72,14 +166,155 @@ class SortedSuccessors {
+    return false;
+}
+
+static bool detectDiamondStructure(BasicBlock* B, ChainWorklist &worklist, IndexSet<BasicBlock*> &done)
```
Ditto about coding styles.

Constellation commented on Source/JavaScriptCore/b3/air/AirOptimizeBlockOrder.cpp:
```diff
@@ -72,14 +166,155 @@ class SortedSuccessors {
+    return true;
+}
+
+static bool optimizeDualSuccessors(BasicBlock* block, ChainWorklist &worklist, IndexSet<BasicBlock*> &done)
```
Ditto.

Constellation commented on Source/JavaScriptCore/b3/air/AirOptimizeBlockOrder.cpp:
```diff
@@ -96,30 +331,49 @@ Vector<BasicBlock*> blocksInOptimizedOrder(Code& code)
+        if (!detectTriangleStructure(block, chainWorklist, done)
+            && !detectDiamondStructure(block, chainWorklist, done)
+            && !optimizeDualSuccessors(block, chainWorklist, done))
+        {
```
Let's make it.
```
    if (!detectTriangleStructure(block, chainWorklist, done)
        && !detectDiamondStructure(block, chainWorklist, done)
        && !optimizeDualSuccessors(block, chainWorklist, done)) {
    ...
```

Constellation commented on Source/JavaScriptCore/b3/air/AirOptimizeBlockOrder.cpp:
```diff
@@ -96,30 +331,49 @@ Vector<BasicBlock*> blocksInOptimizedOrder(Code& code)
+    chainWorklist.append(code.entrypoint(0).block());
+
+    while (BasicBlock* block = chainWorklist.pop(done)) {
+        assert(!done.contains(block));
```
Use `ASSERT`.

Constellation commented on Source/JavaScriptCore/b3/air/AirOptimizeBlockOrder.cpp:
```diff
@@ -57,12 +134,29 @@ class SortedSuccessors {
             [] (BasicBlock* left, BasicBlock* right) {
                 return left->frequency() < right->frequency();
             });
+    }
```
Let's make it a private function.

Constellation commented on Source/JavaScriptCore/b3/air/AirOptimizeBlockOrder.cpp:
```diff
@@ -72,14 +166,155 @@ class SortedSuccessors {
+    return true;
+}
+
+static bool optimizeDualSuccessors(BasicBlock* block, ChainWorklist &worklist, IndexSet<BasicBlock*> &done)
```
Let's name it aligned to the others like,
`detectFallthroughSuccessor`

h4writer replied:
Indeed, it would be better to also use "detectXXX" for this function. Though I'm not sure about "detectFallthroughSuccessor". What do you think about "detectExclusiveSuccessor"? If you don't agree, let me know, and I'll update the patch to use "detectFallthroughSuccessor".
I also added a diagram, like in the other cases, to make it easier to understand.

Constellation reviewed (commented):
Looks good! Some comments.

Constellation reviewed (approved):
Looks good! Before landing, please squash, and let me know when it is ready, I'll put it to a merge-queue!

h4writer committed:
[JSC] Improve BlockOrder by handling finished chains better and optimize specific common block structures.
https://bugs.webkit.org/show_bug.cgi?id=301564

Reviewed by Yusuke Suzuki.

This improves the block order.

1) When there is no successor and the chain is broken,
take a better candidate using FIFO instead of LIFO.
2% more fall-through + 4% decrease in jump lengths

2) Detect common structures and put them in a better order.
Support is included for triangle, diamond, and
two successors (with one having other predecessors).
2.7% more fall-through + 11% decrease in jump lengths

* Source/JavaScriptCore/b3/air/AirOptimizeBlockOrder.cpp:
(JSC::B3::Air::detectTriangleStructure):
(JSC::B3::Air::detectDiamondStructure):
(JSC::B3::Air::detectExclusiveSuccessor):

Canonical link: https://commits.webkit.org/302875@main
